Complex eof

vsilvafelipe opened the pull request:
The ceof and reconstruct_ceof functions were tested in two different datasets (array and DataArray) in two different regions of the ocean. The ceof() returns the eigenvalues, eigenvectors (real and imaginary), amplitudes (real and imaginary), besides the spatial and temporal amplitudes and phases in a DataArray. The reconstruct_ceof() gives the reconstrucion of CEOF modes individually. I am happy to hear and follow any review.

vsilvafelipe committed:
teste ceof

vsilvafelipe committed:
Now, it is possible to input data with NaN values. The function organizes the data as time vs space and subtract the mean field in each coordinate. It returns all the variables related to CEOF in a DataArray.

vsilvafelipe committed:
Update README.md

I added the description for the ceof and reconstruct_ceof modules.

vsilvafelipe committed:
Update __init.py__

I added the ceof, reconstruct_ceof modules in the init file

MariaMigl commented on OceanLab/eof.py:
```diff
@@ -121,3 +121,162 @@ def my_eof_interp(M,nmodes,errmin=1e-15,repmax=None):
+# PERFORM COMPLEX EOF
+#=========================================
+from scipy.signal import hilbert
+import scipy.linalg as la
```
Imports should be placed in the top of each file regardless of the fuction in which they are used

vsilvafelipe replied:
Done

Ryukamusa commented on OceanLab/eof.py:
```diff
@@ -121,3 +121,162 @@ def my_eof_interp(M,nmodes,errmin=1e-15,repmax=None):
+def org_data_ceof(lon, lat, data):
+    data_ceof = np.zeros((len(data), len(lon)*len(lat)))*np.nan
+    k = 0
+    for i in range(len(lat)):
```
This method needs to be refactored. The computation time of a nested loop increases exponentially as the list size increases.

dantecn commented on OceanLab/eof.py:
```diff
@@ -121,3 +121,162 @@ def my_eof_interp(M,nmodes,errmin=1e-15,repmax=None):
+    Translated to Python by Felipe Vilela da Silva @ IMAS, UTas, AU on 15/Mar/2021
+    ==============================================================================
+    Inputs:
+    -> lon: longitudes (array)
```
<pre>
Can you make the inputs following the standards of the other .py? (Maybe it changed after you started this)

It's just capital input/output and the variables below the "U" (3 spaces), with the "=" signed aligned for the larger name:
Example

INPUT:
   Small       = A variable with a small name
   BigNamedOne = A variable with a big name
</pre>

vsilvafelipe replied:
Done

dantecn commented on OceanLab/eof.py:
```diff
@@ -121,3 +121,162 @@ def my_eof_interp(M,nmodes,errmin=1e-15,repmax=None):
+    -> pcs:   First nkp Complex Principal Components or amplitudes [time, nkp]
+    -> t_amp:    Temporal amplitude [time, nkp]
+    -> t_phase:  Temporal phase [time, nkp]    
+    ==============================================================================
```
Please, remove all authorship comments (except for previous credits or paper citations). All contributors from OceanLab Team will be included in a single "contributors" file

vsilvafelipe replied:
Done

vsilvafelipe committed:
Update eof.py

Following Ryukamusa request about "Imports should be placed in the top of each file"

vsilvafelipe committed:
Removed authorship comments

Following dantecn request

vsilvafelipe committed:
Update eof.org_data_ceof()

Now, org_data_ceof works faster and returns a DataArray

vsilvafelipe committed:
Update inputs and outputs labels

Followed @dantecn suggestion.

vsilvafelipe committed:
Update eof.org_data_ceof()

I just realized a typo.

dantecn commented on OceanLab/eof.py:
```diff
@@ -121,3 +125,154 @@ def my_eof_interp(M,nmodes,errmin=1e-15,repmax=None):
+        New modifications:
+        (i)  The function organizes the data as time vs space and subtract the mean field in each coordinate
+        (ii) It returns all the variables related to CEOF in a DataArray
+    ''' 
```
I'm not sure this version control is useful here.
1) github has version control
2) when we merge these commits, it will be the "actual first version" =P

vsilvafelipe replied:
Done

dantecn commented on OceanLab/eof.py:
```diff
@@ -121,3 +125,154 @@ def my_eof_interp(M,nmodes,errmin=1e-15,repmax=None):
+                           coords={'lat':(dims[1], lat), 'lon':(dims[2], lon)})
+    data_ceof = datxarray.stack(lat_lon=("lat", "lon")).data_latlon
+    return data_ceof
+
```
@Ryukamusa @iuryt any idea on how to define these functions that are used only within other functions and used only internally?

Maybe define with the underline "_" at the beginning, as we did for _parallel_computing in the utils module?

`def _org_data_ceof()`
`def _amplitude_phase()`

dantecn replied:
I have to learn where to put my comments o.0

vsilvafelipe replied:
done

vsilvafelipe committed:
Removal of versions in the ceof description

I just followed dantecn suggestion

vsilvafelipe committed:
Update auxiliary functions

"_" at the beginning of auxiliary functions.

vsilvafelipe committed:
Create client within ceof

I added the parallel option to build a standard client within the ceof function. In case the user doesn't want a standard client, he/she can turn the option to False and build the client with other settings.
